store.go

```diff
@@ -0,0 +1,141 @@
+package cbornode
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+
+	block "github.com/ipfs/go-block-format"
+	cid "github.com/ipfs/go-cid"
+	mh "github.com/multiformats/go-multihash"
+	recbor "github.com/polydawn/refmt/cbor"
+	atlas "github.com/polydawn/refmt/obj/atlas"
+	cbg "github.com/whyrusleeping/cbor-gen"
+)
+
+type CborIpldStore interface {
+	Get(ctx context.Context, c cid.Cid, out interface{}) error
+	Put(ctx context.Context, v interface{}) (cid.Cid, error)
+}
+
+type CborIpldBlockstore interface {
+	Get(cid.Cid) (block.Block, error)
+	Put(block.Block) error
+}
+
+type BasicCborIpldStore struct {
+	Blocks CborIpldBlockstore
+	Atlas  *atlas.Atlas
+}
+
+var _ CborIpldStore = &BasicCborIpldStore{}
+
+func NewCborStore(bs CborIpldBlockstore) *BasicCborIpldStore {
+	return &BasicCborIpldStore{Blocks: bs}
+}
+
+func (s *BasicCborIpldStore) Get(ctx context.Context, c cid.Cid, out interface{}) error {
+	blk, err := s.Blocks.Get(c)
+	if err != nil {
+		return err
+	}
+
+	cu, ok := out.(cbg.CBORUnmarshaler)
+	if ok {
+		if err := cu.UnmarshalCBOR(bytes.NewReader(blk.RawData())); err != nil {
+			return NewSerializationError(err)
+		}
+		return nil
+	}
+
+	if s.Atlas == nil {
+		return DecodeInto(blk.RawData(), out)
+	} else {
+		return recbor.UnmarshalAtlased(recbor.DecodeOptions{}, blk.RawData(), out, *s.Atlas)
+	}
+}
+
+type cidProvider interface {
+	Cid() cid.Cid
+}
+
+func (s *BasicCborIpldStore) Put(ctx context.Context, v interface{}) (cid.Cid, error) {
+	mhType := uint64(mh.BLAKE2B_MIN + 31)
+	mhLen := -1
+	codec := uint64(cid.DagCBOR)
```

Comment:
nit, use `cid.Prefix` from the start (if possible)

Reply:
I think we could only do that if there some a guarantee that `v` had a `Cid()` method, unless I am misunderstanding something?